modules/core/src/transitions/gpu-interpolation-transition.js

```diff
@@ -0,0 +1,225 @@
+
+  // get current values of an attribute, clipped/padded to the size of the new buffer
+  _setupBuffers(gl, settings, numInstances) {
+    const {size, normalized} = this.attribute;
+    const precisionMultiplier = this.attribute.doublePrecision ? 2 : 1;
+
+    let toState;
+    if (this.attribute.constant) {
+      toState = new BaseAttribute(gl, {constant: true, value: this.attribute.value, size});
+    } else {
+      toState = new BaseAttribute(gl, {
+        constant: false,
+        buffer: this.attribute.getBuffer(),
+        divisor: 0,
+        size,
+        normalized
+      });
+    }
+
+    const fromState = this.buffer || toState;
+    const toLength =
+      (this.attribute.userData.noAlloc ? this.attribute.value.length : numInstances * size) *
+      precisionMultiplier;
+    const fromLength = this.curLength || toLength;
+
+    // Alternate between two buffers when new transitions start.
+    // Last destination buffer is used as an attribute (from state),
+    // And the other buffer is now the destination buffer.
+    let buffer = this._swapBuffer;
+    this._swapBuffer = this.buffer;
+
+    if (!buffer) {
+      buffer = new Buffer(gl, {
+        data: new Float32Array(toLength),
+        usage: GL.DYNAMIC_COPY
+      });
+    } else if (buffer.getElementCount() < toLength) {
+      // Pad buffers to be the same length with 32-bit floats
+      buffer.reallocate(toLength * 4);
+    }
+
+    this.curLength = toLength;
+    this.attributeInTransition.update({
+      buffer,
+      // Hack: Float64Array is required for double-precision attributes
+      // to generate correct shader attributes
+      value: this.attribute.value
+    });
+
+    // updates the fromState buffer in place
+    padBuffer({
+      fromState,
+      toState,
+      fromLength,
+      toLength,
+      size: size * precisionMultiplier,
+      fromBufferLayout: this.bufferLayout,
+      toBufferLayout: this.attribute.bufferLayout,
+      offset: this.attribute.elementOffset * precisionMultiplier,
+      getData: settings.enter
+    });
+
+    this.bufferLayout = this.attribute.bufferLayout;
+
+    return {toState, buffer, fromState};
+  }
+}
+
+const ATTRIBUTE_MAPPING = {
+  1: 'float',
+  2: 'vec2',
+  3: 'vec3',
+  4: 'vec4'
+};
+
+const vs = `
+#define SHADER_NAME interpolation-transition-vertex-shader
+
+uniform float time;
+attribute ATTRIBUTE_TYPE aFrom;
+attribute ATTRIBUTE_TYPE aTo;
+varying ATTRIBUTE_TYPE vCurrent;
+
+void main(void) {
+  vCurrent = mix(aFrom, aTo, time);
+  gl_Position = vec4(0.0);
+}
+`;
+
+function getShaders(attributeSize) {
+  const attributeType = ATTRIBUTE_MAPPING[attributeSize];
+  return {
+    vs,
+    defines: {
+      ATTRIBUTE_TYPE: attributeType
+    },
+    varyings: ['vCurrent']
+  };
+}
+
+function getBuffers({fromState, toState, buffer}) {
+  return {
+    sourceBuffers: {
+      aFrom: fromState,
+      aTo: toState
+    },
+    feedbackBuffers: {
+      vCurrent: buffer
+    }
+  };
+}
+
+function padBuffer({
```

Comment:
Why is this function moved? Shouldn't all transitions need this?

Reply:
I wasn't sure if it was going to be the same for all transition types. But I think you're right - we can use it in its current form.

Reply:
Moved this back to `attribute-transition-utils` with a comment explaining its use.